src/ed25519/hasher512.cpp

```diff
@@ -127,19 +172,23 @@ namespace libtorrent::aux {
 #elif TORRENT_USE_CRYPTOAPI_SHA_512
 		m_context.reset();
 #elif defined TORRENT_USE_LIBCRYPTO
-		SHA512_Init(&m_context);
+		EVP_DigestInit_ex(m_context, EVP_sha512(), nullptr);
 #else
 		SHA512_init(&m_context);
 #endif
 	}
 
-#if defined TORRENT_USE_LIBGCRYPT
 	hasher512::~hasher512()
 	{
+#if defined TORRENT_USE_LIBGCRYPT
 		gcry_md_close(m_context);
-	}
+#elif defined TORRENT_USE_LIBCRYPTO
+#if OPENSSL_VERSION_NUMBER >= 0x10100000L
+		EVP_MD_CTX_free(m_context);
 #else
-	hasher512::~hasher512() = default;
+		EVP_MD_CTX_destroy(m_context);
```

Comment:
the documentation says:

> EVP_MD_CTX_destroy() cleans up digest context ctx and frees up the space allocated to it, it should be called only on a context created using EVP_MD_CTX_create().

But you may end up calling it on `nullptr` here. I think there needs to be a check, like:

```
if (m_context) EVP_MD_CTX_destroy(m_context);
```

Reply:
I don't see the `_free()` version being documented, but might have the same restriction

Reply:
They do guard against it:
https://github.com/openssl/openssl/blob/d270a6c9ea9f240f653a98c7eb0a5c2134a5d63a/crypto/evp/digest.c#L118-L121

nonetheless I'll still add the check.